### PR TITLE
Remove tokio dependencies

### PR DESCRIPTION
**Pull request by tomaka**
Lots of small changes for #2416 :

- The RPC service now accepts an `Arc<dyn futures::future::Executor<Box<dyn Future>>>` instead of `tokio::runtime::TaskExecutor`, which makes it possible to not depend on tokio anymore.
- The import queue now manually polls the worker if we fail to spawn a future.
- The `Service` itself now works like the import queue: it must be polled, and polling attempts to spawn background tasks. If spawning a task fails, we poll it manually.
- Removes the `TaskExecutor` and `on_exit` from `TelemetryOnConnect`. Grandpa now spawns telemetry futures on the main executor.
- **Biggest breaking change**. Removes the `executor` parameter from various service constructors and components. You can use `service.spawn_task` instead of spawning a task using the executor.
- `import_blocks` now returns a `Future` that performs the operation, instead of blocking.

I unfortunately couldn't add `substrate-rpc` to the CI wasm testing, because [this line](https://github.com/paritytech/jsonrpc/blob/3bee17b525923a14c473be561f677d9f048a7dc7/derive/src/to_client.rs#L115) badly collides with the expectations of `cargo web`. For the initial WASM-browser prototype, I intended to remove the RPC endpoint anyway.

Small note: you can see that I added `.expect("failed to spawn task")` at several locations, and this panic is not proven. However, please note that before this PR we were using `task_executor.spawn()`, which [as you can see here](https://docs.rs/tokio/0.1.21/src/tokio/runtime/threadpool/task_executor.rs.html#56) calls `unwrap`.
In other words, while we now have additional calls to `expect()`, the actual changes of panicking are the same as before, except that before they were hidden.

### CODE REVIEW

File: core/service/src/lib.rs [hunk 1]
```diff
@@ -82,6 +82,14 @@ pub struct Service<Components: components::Components> {
 	keystore: Keystore,
 	exit: ::exit_future::Exit,
 	signal: Option<Signal>,
+	/// Sender for futures that must be spawned as background tasks.
+	to_spawn_tx: mpsc::UnboundedSender<Box<dyn Future<Item = (), Error = ()> + Send>>,
+	/// Receiver for futures that must be spawned as background tasks.
+	to_spawn_rx: Mutex<mpsc::UnboundedReceiver<Box<dyn Future<Item = (), Error = ()> + Send>>>,
```
Comment (bkchr) [inline core/service/src/lib.rs]: Why do we need to put this into a mutex? (below the same question)

File: core/service/src/lib.rs [hunk 2]
```diff
@@ -525,6 +551,50 @@ impl<Components> Service<Components> where Components: components::Components {
...
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		// The user is supposed to poll only one service, so it doesn't matter if we keep this
+		// mutex locked.
+		let mut to_poll = self.to_poll.lock();
+		let mut to_spawn_rx = self.to_spawn_rx.lock();
+
+		while let Ok(Async::Ready(Some(task_to_spawn))) = to_spawn_rx.poll() {
+			let executor = tokio_executor::DefaultExecutor::current();
+			if let Err(err) = executor.execute(task_to_spawn) {
```
Comment (bkchr) [inline core/service/src/lib.rs]: Can we not do this check once?

File: core/service/src/lib.rs [hunk 3]
```diff
@@ -525,6 +551,50 @@ impl<Components> Service<Components> where Components: components::Components {
...
+					target: "service",
+					"Failed to spawn background task: {:?}; falling back to manual polling",
+					err
+				);
+				to_poll.push(err.into_future());
+			}
+		}
+
+		// Polling all the `to_poll` futures.
+		while let Some(pos) = to_poll.iter_mut().position(|t| t.poll().map(|t| t.is_ready()).unwrap_or(true)) {
```
Comment (bkchr) [inline core/service/src/lib.rs]: Just use `retain` here.

Comment (tomaka) [reply core/service/src/lib.rs]: I can't, because `retain` gives you a `&T` and not a `&mut T`.

File: core/service/src/lib.rs (hunk 1 above)
Comment (tomaka) [reply core/service/src/lib.rs]: *sighs* Because we have tests that use an `Arc<Service>`.

File: core/consensus/common/src/import_queue.rs [hunk 4]
```diff
@@ -200,8 +205,21 @@ impl<B: BlockT> ImportQueue<B> for BasicQueue<B> {
 	}
 
 	fn poll_actions(&mut self, link: &mut dyn Link<B>) {
+		// Try to spawn the future in `future_to_spawn`.
 		if let Some(future) = self.future_to_spawn.take() {
-			tokio_executor::spawn(future);
+			if let Err(err) = tokio_executor::DefaultExecutor::current().execute(future) {
+				debug_assert!(self.manual_poll.is_none());
```
[the remark this answers is not in the document]
Comment (tomaka) [reply core/consensus/common/src/import_queue.rs]: I think that's against our "policy" to never panic. `debug_assert!` is ok but `assert!` would have to be proven.

File: core/service/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply core/service/src/lib.rs]: Ahh.. Do you know which one are doing this?

(We should at least file an issue to solve this)

File: core/service/src/lib.rs (hunk 3 above)
Comment (bkchr) [reply core/service/src/lib.rs]: Rust :( I forgot this...

File: core/finality-grandpa/src/lib.rs [hunk 5]
```diff
@@ -723,7 +724,11 @@ pub fn run_grandpa_voter<B, E, Block: BlockT<Hash=H256>, N, RA, SC, X>(
 
 	let voter_work = network_startup.and_then(move |()| voter_work);
 
-	Ok(voter_work.select(on_exit).then(|_| Ok(())))
+	// Make sure that `telemetry_task` doesn't accidentally finish and kill grandpa.
+	let telemetry_task = telemetry_task
+		.then(|_| futures::future::empty::<(), ()>());
```
[the remark this answers is not in the document]
Comment (tomaka) [reply core/finality-grandpa/src/lib.rs]: We want the whole thing to stop if the `voter_task` stops.

File: core/service/src/lib.rs (hunk 2 above)
Comment (tomaka) [reply core/service/src/lib.rs]: No, theoretically the executor could have a limited capacity, and so later tasks can fail to spawn.

File: core/finality-grandpa/src/communication/mod.rs [hunk 6]
```diff
@@ -290,8 +291,11 @@ impl<B: BlockT, N: Network<B>> NetworkBridge<B, N> {
 		let startup_work = futures::future::lazy(move || {
 			// lazily spawn these jobs onto their own tasks. the lazy future has access
 			// to tokio globals, which aren't available outside.
-			tokio::spawn(rebroadcast_job.select(on_exit.clone()).then(|_| Ok(())));
-			tokio::spawn(reporting_job.select(on_exit.clone()).then(|_| Ok(())));
+			let mut executor = tokio_executor::DefaultExecutor::current();
+			executor.spawn(Box::new(rebroadcast_job.select(on_exit.clone()).then(|_| Ok(()))))
+				.expect("failed to spawn task");
```
[the remark this answers is not in the document]
Comment (tomaka) [reply core/finality-grandpa/src/communication/mod.rs]: What do you have in mind? An `error!`? I don't know how to be more explicit than `"failed to spawn task"`.

[the remark this answers is not in the document]
Comment (bkchr) [reply core/finality-grandpa/src/communication/mod.rs]: I would like to have something like `Failed to spawn finality-grandpa rebroadcast job`, to make sure that we find the cause if that fails.

File: core/service/src/lib.rs (hunk 1 above)
Comment (tomaka) [reply core/service/src/lib.rs]: Here: https://github.com/paritytech/substrate/blob/96e17dc57fc7e6744c6c6f824f9c35e31350f709/core/service/test/src/lib.rs#L47-L49

Comment (bkchr) [reply core/service/src/lib.rs]: This looks like it should be a `Box` and not an arc.

**Comment by tomaka** [comment]
I've fixed the connectivity test not working by replacing `shutdown_when_idle` with `shutdown_now`.
Before this PR, the `service` would get dropped before the shutdown. After this PR, since the service has to be polled, we put it in an Arc and spawn it in a separate task.

Unfortunately, just like the network tests, the service test are deeply depending on the fact that we have background threads running, and I don't know how to fix that "properly" without rewriting the tests.
